docs/adr/009-protocols-start-is-an-event.md

```diff
@@ -0,0 +1,33 @@
+# ADR 009 - Model "starting" a protocol as a domain event
+
+```
+Author: @thomaseizinger
+Date: 29/12/2021
+Status: Active
+```
+
+## Summary
+
+Introduce `ContractSetupStarted`, `RolloverStarted` and `CollaborativeSettlementStarted` as domain events.
+
+## Context
+
+[ADR 007](./007-event-sourcing-for-cfds.md) introduced event sourcing for managing a CFD's state.
+Currently, only the completion (with either failure or success) of a protocol like contract-setup is modelled as an event.
+This avoids invalid start up states as the entire protocol state is in-memory until it completes.
+
+Whilst attractive, this has numerous downsides:
+
+1. We cannot easily detect that a certain protocol is already active and thus run the risk of executing it concurrently for the same CFD.
+2. State changes to a CFD are only triggered upon domain events and the UI is only updated on state changes, thus not having such an event introduces additional complexity through manual updates via the projection actor.
+3. We cannot easily detect that a _competing_ protocol is already active.
+   For example, rollover and collaborative settlement are mutually exclusive and should not be executed at the same time.
```

Comment:
```suggestion
   For example, non-collaborative settlement and collaborative settlement are mutually exclusive and should not be executed at the same time.
```

> rollover and collaborative settlement are mutually exclusive

I think this is incorrect. 

Rollover can be triggered while collaboratively settlement and collaboratively settling can be triggered while rolling over.
Rollover does not alter the lock transaction, so agreeing to spend from it in collaborative settlement should always work regardless of a potential ongoing rollover. The only side effect might be that rollover reports a failure when trying to produce the event upon completion, because the CFD is already final. (Rollover completing while an ongoing collaborative settlement should not have any affect, because rollover does not make the cfd final.)

That said: I think we should not block collaborative settlement if there is an ongoing rollover! The user should always be able to trigger a collaborative settlement, unless the CFD is already closed non-collaboratively. Did you mean non-collaborative settlement instead of rollover here?

Reply:
No, this was actually on purpose!

1. I think non-collab settlement should never be blocked on anything.
2. I thought that from a business perspective, rolling over a contract is the opposite of settling it so they should not happen concurrently. You can settle right after or before rollover but not while it is active. I don't mind how we implement it but restricting this made sense to me. On BitMex, rollover is atomic so not really a concern but I doubt you'd be able to close / sell your position if you were in the middle of being charged interest.